Add customElements.getName method

rniwa opened the pull request:
#### e2748d7c56a29d295d71552e3d999d7b51ca037e
<pre>
Add customElements.getName method
<a href="https://bugs.webkit.org/show_bug.cgi?id=255778">https://bugs.webkit.org/show_bug.cgi?id=255778</a>

Reviewed by Alexey Shvayka and Chris Dumez.

Implement <a href="https://github.com/whatwg/html/pull/9195">https://github.com/whatwg/html/pull/9195</a> with test cases added in
<a href="https://github.com/web-platform-tests/wpt/pull/39640.">https://github.com/web-platform-tests/wpt/pull/39640.</a>

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry.html:
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::getName): Added.
* Source/WebCore/dom/CustomElementRegistry.h:
* Source/WebCore/dom/CustomElementRegistry.idl:

Canonical link: <a href="https://commits.webkit.org/263281@main">https://commits.webkit.org/263281@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f21fc1a48bc8cec2c79a2e607a9f5cdc7084c439

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4117 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4226 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5576 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4359 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4104 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4324 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4191 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4586 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4173 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4351 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3713 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5569 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1848 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3690 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5863 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3675 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3753 "1 flakes 3 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5262 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4150 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3360 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3664 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3688 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/7801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/477 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3943 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shvaikalesh commented on Source/WebCore/bindings/js/JSCustomElementRegistryCustom.cpp:
```diff
@@ -202,6 +202,26 @@ JSValue JSCustomElementRegistry::define(JSGlobalObject& lexicalGlobalObject, Cal
+        return throwException(&lexicalGlobalObject, scope, createNotEnoughArgumentsError(&lexicalGlobalObject));
+
+    auto constructorValue = callFrame.uncheckedArgument(0);
+    if (!constructorValue.isConstructor())
```
Do we absolutely need to check it's specifically constructor here? Like if it isn't, `findInterface` will return `nullptr`. We may consider simplifying this check to `wrapped().findInterface(callFrame.uncheckedArgument(0).getObject())` or something.

rniwa replied:
Good point. We can simplify that.

shvaikalesh commented on Source/WebCore/dom/CustomElementRegistry.idl:
```diff
@@ -31,6 +31,7 @@
 ] interface CustomElementRegistry {
     [CEReactions, Custom] undefined define(DOMString name, Function constructor);
     any get([AtomString] DOMString name);
+    [Custom] DOMString? getName(any constructor);
```
```suggestion
    [Custom] DOMString? getName(Function constructor);
```

rniwa replied:
Hm... this results in an error like this:
```
Could NOT find IDL file for interface "Function"!
Could NOT find IDL file for interface "Function"!

cdumez commented on Source/WebCore/dom/CustomElementRegistry.idl:
```diff
@@ -31,6 +31,7 @@
 ] interface CustomElementRegistry {
     [CEReactions, Custom] undefined define(DOMString name, Function constructor);
     any get([AtomString] DOMString name);
+    [CallWith=CurrentGlobalObject] any getName(any constructor);
```
I would be nice to add a FIXME comment explaining that the return value should be `(DOMString or undefined)` but that our bindings generator doesn't support this yet.

rniwa replied:
Sure. I'd add a FIXME with a reference to https://bugs.webkit.org/show_bug.cgi?id=232734

cdumez reviewed (approved):
LGTM with nit.

rniwa commented on Source/WebCore/dom/CustomElementRegistry.cpp:
```diff
@@ -128,6 +128,17 @@ JSC::JSValue CustomElementRegistry::get(const AtomString& name)
+    JSC::JSObject* constructor = constructorValue.getObject();
+    if (!constructor)
+        return JSC::jsUndefined();
+    auto* interface = findInterface(constructor);
```
Can't use the name "interface" on Windows. It's why WinCairo is failing to build.
This comes up every time I touch this file. It's annoying as hell...

rniwa committed:
Add customElements.getName method
https://bugs.webkit.org/show_bug.cgi?id=255778

Reviewed by Alexey Shvayka and Chris Dumez.

Implement https://github.com/whatwg/html/pull/9195 with test cases added in
https://github.com/web-platform-tests/wpt/pull/39640.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry.html:
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::getName): Added.
* Source/WebCore/dom/CustomElementRegistry.h:
* Source/WebCore/dom/CustomElementRegistry.idl:

Canonical link: https://commits.webkit.org/263281@main
